From 3a3cca4d3a6a4b763b5e88b36ec77673c78e020c Mon Sep 17 00:00:00 2001
From: Greg Steuck <greg@nest.cx>
Date: Wed, 14 Sep 2022 06:52:10 -0700
Subject: [PATCH] Revert "driver: don't actually merge objects when ar -L
 works"

This reverts commit 112e4f9c9c299b460e37a60d8f8d8693aa6ab06a.

The commit triggers a build problem on OpenBSD:

===> Command failed with error code: 1
ld.lld: error: .../x86_64-openbsd-ghc-9.5.20220824/libHSdirectory-1.3.7.1-ghc9.5.20220824.so: undefined reference to unixzm2zi7zi2zi2_SystemziPosixziFiles_setFileMode1_info [--no-allow-shlib-undefined]
ld.lld: error: .../x86_64-openbsd-ghc-9.5.20220824/libHSdirectory-1.3.7.1-ghc9.5.20220824.so: undefined reference to unixzm2zi7zi2zi2_SystemziPosixziFiles_setFileMode1_closure [--no-allow-shlib-undefined]
...
ld.lld: error: too many errors emitted, stopping now (use -error-limit=0 to see all errors)
clang: error: linker command failed with exit code 1 (use -v to see invocation)
`clang' failed in phase `Linker'. (Exit code: 1)
Error when running Shake build system:
  at action, called at src/Rules.hs:38:19 in main:Rules
  at need, called at src/Rules.hs:60:5 in main:Rules
* Depends on: _validatebuild/stage1/bin/hpc
  at cmd', called at src/Builder.hs:386:23 in main:Builder
  at cmdArgs, called at src/Builder.hs:539:8 in main:Builder
  at cmdArgs, called at src/Builder.hs:563:18 in main:Builder
  at cmdArgs, called at src/Builder.hs:563:18 in main:Builder
  at error, called at src/Builder.hs:608:13 in main:Builder
* Raised the exception:
Command failed
Build failed.

112e4f9c9c299b460e37a60d8f8d8693aa6ab06a is the first bad commit
commit 112e4f9c9c299b460e37a60d8f8d8693aa6ab06a
Author: Cheng Shao <astrohavoc@gmail.com>
Date:   Tue Aug 23 14:45:58 2022 +0000
    driver: don't actually merge objects when ar -L works
 compiler/GHC/Driver/Pipeline/Execute.hs | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
bisect found first bad commit
---
 compiler/GHC/Driver/Pipeline/Execute.hs | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/compiler/GHC/Driver/Pipeline/Execute.hs b/compiler/GHC/Driver/Pipeline/Execute.hs
index b3737dc7e89..4775e6971c3 100644
--- a/compiler/GHC/Driver/Pipeline/Execute.hs
+++ b/compiler/GHC/Driver/Pipeline/Execute.hs
@@ -1182,7 +1182,7 @@ via gcc.
 -- | See Note [Object merging].
 joinObjectFiles :: HscEnv -> [FilePath] -> FilePath -> IO ()
 joinObjectFiles hsc_env o_files output_fn
-  | can_merge_objs && not dashLSupported = do
+  | can_merge_objs = do
   let toolSettings' = toolSettings dflags
       ldIsGnuLd = toolSettings_ldIsGnuLd toolSettings'
       ld_r args = GHC.SysTools.runMergeObjects (hsc_logger hsc_env) (hsc_tmpfs hsc_env) (hsc_dflags hsc_env) (
-- 
GitLab

